src/optint.ml

```diff
@@ -0,0 +1,49 @@
+module type Immediate = sig
+  type t [@@immediate]
+end
+
+module type Boxed = sig
+  type t
+end
+
+module Immediate64 (Immediate : Immediate) (Boxed : Boxed) = struct
+  type t [@@immediate64]
+  type 'a repr = Immediate : Immediate.t repr | Boxed : Boxed.t repr
+
+  let repr =
+    (* For soundness of the [@@immediate64] annotation, we ensure to use the
+       boxed representation only when not on 64-bit platforms, but we need to
+       use The Force to convince the type system of this fact. *)
+    match Sys.word_size with
+    | 64 -> (Obj.magic Immediate : t repr)
+    | 32 -> (Obj.magic Boxed : t repr)
```

Comment:
😕 I still prefer `(select ...)` than that... sorry.

Reply:
I'm not sure what solution you're imagining w/ `(select ...)` without first needing a [more powerful `select` stanza](https://github.com/ocaml/dune/issues/397).

Reply:
Even with the ability to select between multiple _interfaces_ the library would end up needing to expose that conditional choice in an observable way, and this can break downstream code. (This was also true of the Dune rules I had before.)

So I think some sort of `Obj.magic` is going to be necessary to have this annotation.

Reply:
> to expose that conditional choice in an observable way

I'm not sure to understand why we need that. In my mind, the end user should not introspect how `optint` chosen its implementation - and if the user really need to introspect that, he can do that with `Sys.word_size`. I suspect that you need such observable value on tests/fuzzers but, again, it's a really limited scope.

More generally, `Obj.magic` is not the right solution 😕 . It can blows up our stack at a deep level.

Reply:
It's precisely that we _don't_ want that that's the problem. With an implementation in the build system, the `[@immediate64]` annotation is added conditionally according to the platform, and so downstream code that _also_ uses `[@immediate64]` will break when Optint supplies the abstract type `t` rather than the annotated one. Generally, it seems to me to be a bad idea to use the build system to select between interfaces: it breaks the abstraction. I'm not a fan of `Obj.magic` either, but I don't see any other way to avoid this (and I think it's fairly well-scoped here to be safe).

On a not-completely-unrelated note, I think it _is_ important for users to introspect at runtime on the representation, so that we can use `int63` for file offsets in C bindings (as in [this PR](https://github.com/mirage/index/pull/286)) w/o needing to allocate on 64-bit platforms. That can be done with type equality witnesses though, so is type-safe.

Reply:
> It's precisely that we don't want that that's the problem. With an implementation in the build system, the [@immediate64] annotation is added conditionally according to the platform, and so downstream code that also uses [@immediate64] will break when Optint supplies the abstract type t rather than the annotated one.

I understand your point but I was believed that C stubs (in particular) should not widely uses `optint`. I think, we possibly mix both solution between `(select ...)` (or whatever do the choice such as our own executable) and your GADT. What do you think about an essentialization of the choice between `Immediate` and `Boxed` via a pre-process (as we do now) and a derivation of `optint.ml` and `int63.ml` from this generated code-source?

Reply:
I think any selection at the build-system level is concerning because it's not abstracting away the selection. e.g. the following would compile on 32-bit platforms and be broken on 64-bit platforms:

```ocaml
module Offset : sig
  type t [@@immediate64]
end = struct
  type t = Int63.t
end
```

Am I right in thinking you'd rather have this breakage than use `Obj.magic` to avoid it?

Reply:
> I understand your point but I was believed that C stubs (in particular) should not widely uses `optint`.

That would limit the usefulness of this library for systems code quite a bit, I think. e.g. in Index, if we can't use `int63` for C stubs then we may as well suffer the allocation upfront and use `int64` for file offsets everywhere. I think this problem is orthogonal to the question of the `[@immediate64]` annotation, since it's a run-time observation of the type.